[LVFM-1067] - use bolos-enclave's aes-gcm

pcollet-ledger opened the pull request.

pcollet-ledger committed:
Adding PIC macro to algorithm names search

pcollet-ledger committed:
Plug bolos RNG to noise RNG interface

pcollet-ledger committed:
Adding Bolos backend

pcollet-ledger committed:
Removing compilation warnings

pcollet-ledger committed:
Removing blake2b and blake2s implementation not based on bolos cx_hash syscall

pandouard committed:
Add noise_cipherstate_set_nonce_for_psd

pandouard committed:
Correctly allocate memory

pcollet-ledger committed:
Change backend name to bolos-vaultapp

pcollet-ledger committed:
Put specific modifications for vaultApp (out of backend), in preprocessing #ifdeg

pcollet-ledger committed:
Add ifdef (to rebase -i)

pcollet-ledger committed:
Fix bad #ifdef

pcollet-ledger committed:
NOISE_USE_BOLOS_VAULTAPP_BACKEND become USE_BOLOS_VAULTAPP_BACKEND

pcollet-ledger committed:
Remove commented lines

pcollet-ledger committed:
Bring back aes-gcm from vaultapp to noise-c' bolos backend

pcollet-ledger committed:
Add vault-hsm backend

pcollet-ledger committed:
Use same cipher-aesgcm as bolos-vaultapp backend

pcollet-ledger committed:
Add set_aes_gcm_functions get_gcm_state_size in vault-hsm backend

pcollet-ledger committed:
Remove os.h include vault-hsm cipher-aes-gcm

pcollet-ledger committed:
Add usage of noise_aesgcm_new_vaulthsm function in internal.c

pcollet-ledger committed:
Add -DUSE_BOLOS_VAULTHSM_BACKEND=1 when compiling VAULTHSM backend

pcollet-ledger committed:
Add aes-gcm implementation for vault-hsm backend

pcollet-ledger committed:
Unify bolos-enclave and bolos aes calls in aes-gcm.c

pcollet-ledger committed:
Correctly wipe aes key

pcollet-ledger committed:
Move aes-gcm implementation in common directory

pcollet-ledger committed:
Use package-config to bring bolos dependencies

hcleonis commented on src/backend/bolos-vaultapp/cipher-aesgcm.c:
```diff
@@ -0,0 +1,147 @@
+        (buf)[4] = (uint8_t)(_value >> 24); \
+        (buf)[5] = (uint8_t)(_value >> 16); \
+        (buf)[6] = (uint8_t)(_value >> 8); \
+        (buf)[7] = (uint8_t)_value; \
```
Can't you bitwise OR-combine all these byte shift operations into a single assignment expression? It should also helps the compiler in optimizing that operation.

pandouard replied:
We do not touch this part

hcleonis commented on src/backend/bolos-vaultapp/cipher-aesgcm.c:
```diff
@@ -0,0 +1,147 @@
+    if (aes_gcm_ae(st->k, st->k_len, st->iv, st->iv_len, tmp, len, ad, ad_len,
+    		   data /* cipher */, data + len /* tag */) < 0) {
+
+      free(tmp);
```
You might want to cleanup or ramdomize `tmp` before freeing it so as to leave no intermediate cyyptographic result in RAM.

pandouard replied:
Confirmed by Donjon, this need to be done!

pcollet-ledger replied:
done

hcleonis commented on src/backend/bolos-vaultapp/cipher-aesgcm.c:
```diff
@@ -0,0 +1,147 @@
+    }
+
+
+    free(tmp);
```
Same here.

pcollet-ledger replied:
done

hcleonis commented on src/backend/bolos-vaultapp/cipher-aesgcm.c:
```diff
@@ -0,0 +1,147 @@
+    noise_aesgcm_setup_iv(st);
+    if (aes_gcm_ad(st->k, st->k_len, st->iv, st->iv_len, tmp, len, ad, ad_len,
+    		   data + len, data) < 0) {
+      free(tmp);
```
Same here

pcollet-ledger replied:
done

hcleonis commented on src/backend/bolos-vaultapp/cipher-aesgcm.c:
```diff
@@ -0,0 +1,147 @@
+      return NOISE_ERROR_MAC_FAILURE;
+    }
+
+    free(tmp);
```
Same here

pcollet-ledger replied:
done

hcleonis commented on src/backend/bolos-vaultapp/cipher-aesgcm.c:
```diff
@@ -0,0 +1,147 @@
+    return NOISE_ERROR_NONE;
+}
+
+NoiseCipherState *noise_aesgcm_new_bolos(void)
```
Function name is a bit confusing. I'd propose 
```c
NoiseCipherState *noise_bolos_aesgcm_new(void)
```
or 
```c
NoiseCipherState *bolos_noise_aesgcm_new(void)
```

pcollet-ledger replied:
If you check `internal.c` this is the way they are named. 
But bolos is not correct, I replaced it by `noise_aesgcm_new_vaultapp`

hcleonis commented on src/backend/bolos-vaultapp/dh-curve25519.c:
```diff
@@ -0,0 +1,181 @@
+    v[i] = v[j];
+    v[j] = t;
+    i++;
+    j--;
```
`i` (resp. `j`) is already incremented (resp. decremented) in the for loop. So it seems that this code is swapping 1 byte out of 2 instead of swapping all the bytes. Is this the intended behaviour?

pandouard replied:
Yes it is the expected behaviour

hcleonis commented on src/backend/bolos-vaultapp/dh-curve25519.c:
```diff
@@ -0,0 +1,181 @@
+  NoiseCurve25519State *st = (NoiseCurve25519State *)state;
+
+  noise_rand_bytes(st->private_key, sizeof(st->private_key));
+  noise_curve25519_set_keypair_private(state, st->private_key);
```
Is it OK not to check the return value of this function (it returns an `int`)?

pcollet-ledger replied:
Return code is now verified. Done.

hcleonis commented on src/backend/bolos-vaultapp/dh-curve25519.c:
```diff
@@ -0,0 +1,181 @@
+{
+  /* Check that the public key actually corresponds to the private key */
+  NoiseCurve25519State *st = (NoiseCurve25519State *)state;
+  NoiseCurve25519State temp;
```
If this code is intended to be compiled with the vaultApp, this would declare `65 bytes + sizeof(struct NoiseDHState_s)` on the stack, which is not that big.  
You might want to declare it as an overlapped static global instead, or to allocate it on the heap instead, depending on the vault-python-tests results with speculos or a real device.

hcleonis commented on src/backend/bolos-vaultapp/dh-curve25519.c:
```diff
@@ -0,0 +1,181 @@
+{
+  NoiseCurve25519State *st = (NoiseCurve25519State *)state;
+  uint8_t tmp_public[sizeof(C_Curve25519_G)];
+  uint8_t tmp_private[32];
```
Same here. And since this function is called from `noise_curve25519_set_keypair()`, this makes `130 bytes + sizeof(struct NoiseDHState_s)` reserved on the stack. for a call to `noise_curve25519_set_keypair()`.

hcleonis commented on src/backend/bolos-vaultapp/dh-curve25519.c:
```diff
@@ -0,0 +1,181 @@
+  uint8_t tmp_private[32];
+
+  memcpy(st->private_key, private_key, 32);
+  memcpy(tmp_public, C_Curve25519_G, sizeof(C_Curve25519_G)); // little endian
```
If this code is compiled with the vaultApp, then the `C_Curve25519_G` array being declared as `const` is stored in the read-only NVM segment of the app, and accessing it requires the us tof the `PIC()` macro.
```c
 memcpy(tmp_public, PIC(C_Curve25519_G), sizeof(C_Curve25519_G));
```

hcleonis commented on src/backend/bolos-vaultapp/dh-curve25519.c:
```diff
@@ -0,0 +1,181 @@
+}
+
+static int noise_curve25519_copy
+    (NoiseDHState *state, const NoiseDHState *from, const NoiseDHState *other)
```
`other` pointer not used at all, is that normal?
And if that's normal, shan't we check that it is set to NULL and return an error if it's not?

pcollet-ledger replied:
This is copy pasted code from Noise.

hcleonis commented on src/backend/bolos-vaultapp/dh-curve25519.c:
```diff
@@ -0,0 +1,181 @@
+{
+    NoiseCurve25519State *st = (NoiseCurve25519State *)state;
+    const NoiseCurve25519State *from_st = (const NoiseCurve25519State *)from;
+    memcpy(st->private_key, from_st->private_key, sizeof(st->private_key));
```
Are you sure you don't need to check that `from` & `st` pointers are not `NULL` before de-referencing them (e.g. because caller can never set `NULL` values for these pointer)?

pcollet-ledger replied:
This is copy pasted code from noise-c.

hcleonis commented on src/backend/bolos-vaultapp/dh-curve25519.c:
```diff
@@ -0,0 +1,181 @@
+
+  memcpy(shared_key, tmp + 1, 32);
+
+  return NOISE_ERROR_NONE;
```
`tmp` contains a shared key (the shared channel encryption key?), so it should be randomized before returning from this function.

pcollet-ledger replied:
Done

hcleonis commented on src/backend/bolos-vaultapp/dh-curve25519.c:
```diff
@@ -0,0 +1,181 @@
+  be2le(tmp + 1, 32);
+  tmp[0] = 0x02;
+
+  uint8_t tmp_private[32];
```
65 bytes of temp buffers allocated on the stack. We might want to `malloc()` them instead.

hcleonis commented on src/backend/vault-common/aes-gcm.c:
```diff
@@ -0,0 +1,429 @@
+		res |= aa[i] ^ bb[i];
+
+	return res;
+}
```
We might want to deactivate compiler optimizations here to ensure that the generated cod will indeed perform a constant-time comparison. To be checked w/ generated assembly code.

pcollet-ledger replied:
Copy-pasted code. I don't know what to change ^^ .

hcleonis commented on src/backend/vault-common/aes-gcm.c:
```diff
@@ -0,0 +1,429 @@
+    return -1;
+  return 0;
+#else /* USE_BOLOS_VAULTHSM_BACKEND */
+  memset(&aes, 0, sizeof(*aes));
```
Shouldn't keys be randomized instead to clear them from memory?
Instead, one should use the explicit_bzero function for the vaultApp

pcollet-ledger replied:
`!AES_INIT_KEY(zero_key, MAX_SYMMETRIC_KEY_LEN, aes)` set the key to zero. I can't tell myself if explicit_bzero will wipe the key or not

hcleonis commented on src/backend/vault-common/aes-gcm.c:
```diff
@@ -0,0 +1,429 @@
+{
+	uint8_t H[AES_BLOCK_SIZE];
+	uint8_t J0[AES_BLOCK_SIZE];
+	uint8_t S[16], T[16];
```
Total of 64 byets allocated on the stack. Careful if this code is alsi common to the vaultApp.

pcollet-ledger replied:
This will be done in the future if we need more stack

hcleonis commented on src/backend/vault-hsm/cipher-aesgcm.c:
```diff
@@ -0,0 +1,160 @@
+        (buf)[5] = (uint8_t)(_value >> 16); \
+        (buf)[6] = (uint8_t)(_value >> 8); \
+        (buf)[7] = (uint8_t)_value; \
+    } while (0)
```
Why not bitwise OR-combine all these byte shifts in a single expression?

hcleonis commented on src/backend/vault-hsm/cipher-aesgcm.c:
```diff
@@ -0,0 +1,160 @@
+    		   data /* cipher */, data + len /* tag */) < 0) {
+
+      free(tmp);
+      return NOISE_ERROR_MAC_FAILURE;
```
We might want to clean intermediate crypto results before returning here?

pcollet-ledger replied:
done

hcleonis commented on src/backend/vault-hsm/cipher-aesgcm.c:
```diff
@@ -0,0 +1,160 @@
+      return NOISE_ERROR_MAC_FAILURE;
+    }
+
+    free(tmp);
```
Same here?

pcollet-ledger replied:
Done

hcleonis commented on src/backend/vault-hsm/cipher-aesgcm.c:
```diff
@@ -0,0 +1,160 @@
+
+    free(tmp);
+
+    return NOISE_ERROR_NONE;
```
Same as above in this fucntion also?

pcollet-ledger replied:
done

hcleonis commented on src/backend/vault-hsm/cipher-chachapoly.c:
```diff
@@ -0,0 +1,158 @@
+        (buf)[4] = (uint8_t)((value) >> 32); \
+        (buf)[5] = (uint8_t)((value) >> 40); \
+        (buf)[6] = (uint8_t)((value) >> 48); \
+        (buf)[7] = (uint8_t)((value) >> 56); \
```
bitwise-OR?

hcleonis commented on src/backend/vault-hsm/dh-curve448.c:
```diff
@@ -0,0 +1,126 @@
+}
+
+static int noise_curve448_copy
+    (NoiseDHState *state, const NoiseDHState *from, const NoiseDHState *other)
```
Since `other` is not used, shan't we check that it is set to NULL and return an error if it's not?

hcleonis commented on src/backend/vault-hsm/dh-curve25519.c:
```diff
@@ -0,0 +1,131 @@
+}
+
+static int noise_curve25519_copy
+    (NoiseDHState *state, const NoiseDHState *from, const NoiseDHState *other)
```
Since `other` is not used, shan't we check that it is set to NULL and return an error if it's not?

pcollet-ledger replied:
Copy pasted noise-c code

hcleonis commented on src/backend/vault-hsm/dh-curve25519.c:
```diff
@@ -0,0 +1,131 @@
+} NoiseCurve25519State;
+
+static int noise_curve25519_generate_keypair
+    (NoiseDHState *state, const NoiseDHState *other)
```
Since `other` is not used, shan't we check that it is set to NULL and return an error if it's not?

hcleonis commented on src/backend/vault-hsm/dh-curve448.c:
```diff
@@ -0,0 +1,126 @@
+static uint8_t const basepoint[56] = {5};
+
+static int noise_curve448_generate_keypair
+    (NoiseDHState *state, const NoiseDHState *other)
```
Since `other` is not used, shan't we check that it is set to NULL and return an error if it's not?

hcleonis commented on src/backend/vault-hsm/dh-newhope.c:
```diff
@@ -0,0 +1,177 @@
+        /* Generating the keypair for Bob relative to Alice's parameters */
+        if (!os || os->parent.key_type == NOISE_KEY_TYPE_NO_KEY)
+            return NOISE_ERROR_INVALID_STATE;
+        noise_rand_bytes(st->random_data, st->parent.private_key_len);
```
Why do you ensure that the `other` pointer is not `NULL` before de-referencing it (through the `if (!os||...)` statement :arrow_up:), but you don't check the validity of the `st` pointer before accessing it? Is that latter guaranteed by caller to be non-NULL?

hcleonis commented on src/crypto/blake2/blake2-endian.h:
```diff
@@ -44,6 +44,8 @@
 #define __LITTLE_ENDIAN __DARWIN_LITTLE_ENDIAN
 #endif
+#elif defined(LEDGER_VAULTHSM)
+#include <machine/endian.h>
```
Is that OK in native mode?

pcollet-ledger replied:
Cannot test it but it compiles So I guess yes :D we will test in native once the PR is merged

hcleonis commented on src/crypto/ghash/ghash.c:
```diff
@@ -43,6 +43,9 @@
 #endif
+#elif defined(LEDGER_VAULTHSM)
+#define __LITTLE_ENDIAN 1234
+#define __BYTE_ORDER __LITTLE_ENDIAN
```
Same question as :arrow_up: here

hcleonis commented on src/crypto/goldilocks/src/include/word.h:
```diff
@@ -26,9 +26,13 @@
 #endif
+#ifdef LEDGER_VAULTHSM
+#include <machine/endian.h>
+#else
```
And here

pcollet-ledger replied:
Same

hcleonis commented on src/protocol/handshakestate.c:
```diff
@@ -1779,4 +1779,17 @@ int noise_handshakestate_get_handshake_hash
+
+NoiseDHState *noise_handshakestate_get_ephemeral_dh(NoiseHandshakeState *state)
+{
+    return state ? state->dh_local_ephemeral : 0;
```
You should use NULL (which is generally defined as `(void*)0` (so a real pointer) instead of 0 to silence a potential conversion warning.

hcleonis commented on src/protocol/rand_os.c:
```diff
@@ -99,9 +105,16 @@ void noise_rand_bytes(void *bytes, size_t size)
+      return;
+    fprintf(stderr, "bls_rng failed\n");
+#elif defined(LEDGER_VAULTAPP)
+    cx_rng(bytes, size);
```
No error handling for vaultApp?

pandouard replied:
Apparently not possible to detect if the rng failed

hcleonis reviewed (changes requested):
I'm suspecting that all the files stored in `src/backend/bolos-vaultapp` are compiled with the vaultApp, right? So since they use bolos primitives which are not publicly exposed, I'm wondering  if they shouldn't be stored on git-orange instead? Donjon opinion required here. 

Same question for the files stored in `src/vault-common` which I suppose are buuilt both with the vaultApp & with the HSM scripts? Same also for files stored in `src/vault-hsm`.

pcollet-ledger committed:
Plug bls_hash primitive for sha256 and sha512 calls in vaulthsm backend

pcollet-ledger committed:
Remove unused files from vaulthsm backend

pandouard committed:
Correct a memory issue in aes_wipe_key

For the vaultApp target, an incorrect address was passed to clear the
aes buffer.

pandouard commented on src/backend/bolos-vaultapp/cipher-aesgcm.c:
```diff
@@ -0,0 +1,147 @@
+    NoiseAESGCMState *st = (NoiseAESGCMState *)state;
+    uint8_t *tmp = NULL;
+
+    if (len > 0) {
```
What happens if len == 0? 
Is it possible?

pandouard replied:
Yes it is possible and without that check the code is unusable.

pcollet-ledger committed:
wipe aes_gcm tmp context before freeing

pcollet-ledger committed:
explicit_bzero tmp in dh-curve25519 temp buffer

pcollet-ledger committed:
rename noise_aesgcm_new_bolos into noise_aesgcm_new_vaultapp

pcollet-ledger committed:
verify cx_ecfp_scalar_mult return code correctly in noise_curve25519_set_keypair_private

pandouard committed:
Fix off by one when handling in curve25519 public keys

pandouard committed:
noise_curve25519_set_keypair: fix wrong size in pubkey verification

sizeof was used on a pointer, hence only the first 4 bytes of the
public key were compared.

pcollet-ledger committed:
Exiting when bls_rng fail

pcollet-ledger committed:
Fix vault-hsm sha256/512 warnings
